### PR TITLE
chore(react-client): introduces new types to checkout order / session

### PR DESCRIPTION
**Pull request by ghost**
## Description

This PR adds new types for the checkout order and checkout session. These types will be necessary for a new checkout feature: new impersonation flow to post an order

## Checklist

- [x] The commit message follows our guidelines
- [x] Tests for the respective changes have been added
- [x] The code is commented, particularly in hard-to-understand areas
- [x] The labels and/or milestones were added

## Disclaimer

By sending us your contributions, you are agreeing that your contribution is made subject to the terms of our [Contributor Ownership Statement](https://github.com/Farfetch/.github/blob/master/COS.md)

### CODE REVIEW

File: packages/client/src/checkout/types/checkoutOrder.types.ts [hunk 1]
```diff
@@ -78,4 +78,7 @@ export type CheckoutOrder = {
   shippingMode: ShippingMode;
   paymentIntentId?: PaymentIntent['id'];
   promotionEvaluationId?: PromotionEvaluationId;
+  customerId?: number;
```
Comment (anapim) [inline packages/client/src/checkout/types/checkoutOrder.types.ts:81]: I don't think these types belong here. From what I understood they're used on the request to create the order and not the response.

File: packages/client/src/checkout/types/checkoutSession.types.ts [hunk 2]
```diff
@@ -53,4 +53,6 @@ export type CheckoutSession = {
   promotionEvaluationId?: string;
   totalCredit?: number;
   promocode?: string;
+  customerId?: number;
```
Comment (anapim) [inline packages/client/src/checkout/types/checkoutSession.types.ts:56]: Same here.

**Review by anapim — CHANGES_REQUESTED** [review]

File: packages/client/src/checkout/types/checkoutOrder.types.ts (hunk 1 above)
Comment (hichico) [reply packages/client/src/checkout/types/checkoutOrder.types.ts]: Changed to the post request. Let me know if this is what was intended.

Comment (anapim) [reply packages/client/src/checkout/types/checkoutOrder.types.ts]: Now it looks ok!
But on further investigation I think I've missed something before. `customerId` should be part of both the `checkoutOrder` and `checkoutSession` (as you had previously but **only** for `customerId`)

File: packages/client/src/checkout/types/postCheckoutOrder.types.ts [hunk 3]
```diff
@@ -18,6 +18,9 @@ export type PostCheckoutOrderItem = {
   quantity: number;
   customAttributes: string;
   productAggregatorId: number;
+  tenantId?: number;
+  customerId?: number;
+  isGuestUser?: boolean;
```
Comment (anapim) [inline packages/client/src/checkout/types/postCheckoutOrder.types.ts:23]: Move to `PostCheckoutOrderdata`

**Review by anapim — CHANGES_REQUESTED** [review]

**Review by anapim — APPROVED** [review]

File: packages/client/src/checkout/types/postCheckoutOrder.types.ts [hunk 4]
```diff
@@ -9,6 +9,9 @@ export type PostCheckoutOrderData = {
   guestUserEmail?: string;
   shippingMode?: ShippingMode;
   metadata?: PostCheckoutOrderMetadata;
+  tenantId?: number;
+  customerId?: number;
+  isGuestUser?: boolean;
```
Comment (boliveira) [inline packages/client/src/checkout/types/postCheckoutOrder.types.ts:14]: Question: Does the server validate if invalid combination of parameters are sent? For example, I could send `guestUserEmail` with a value but set `isGuestUser` to false. What will happen in that case? If it is invalid, then I think this type should be divided in 2: One when `guestUserEmail` is a string which will restrict the type of `isGuestUser` to `true | undefined` and another where `guestUserEmail` is undefined and `isGuestUser` can be then `boolean | undefined`.

Comment (boliveira) [inline packages/client/src/checkout/types/postCheckoutOrder.types.ts:14]: Question: Does the server validate if invalid combination of parameters are sent? For example, I could send `guestUserEmail` with a value but set `isGuestUser` to false. What will happen in that case? If it is invalid, then I think this type should be divided in 2: One when `guestUserEmail` is a string which will restrict the type of `isGuestUser` to `true | undefined` and another where `guestUserEmail` is undefined and `isGuestUser` can be then `boolean | undefined`.

File: packages/client/src/checkout/types/postCheckoutSession.types.ts [hunk 5]
```diff
@@ -11,6 +11,9 @@ export type PostCheckoutSessionData = {
   metadata?: CheckoutSessionMetadata;
   successUrl: string;
   cancelUrl: string;
+  tenantId?: number;
+  customerId?: number;
+  isGuestUser?: boolean;
```
Comment (boliveira) [inline packages/client/src/checkout/types/postCheckoutSession.types.ts:16]: Same as https://github.com/Farfetch/blackout/pull/933/files#r1413557582

Comment (boliveira) [inline packages/client/src/checkout/types/postCheckoutSession.types.ts:16]: Same as https://github.com/Farfetch/blackout/pull/933/files#r1413557582

**Comment by dnamorim** [comment]
This PR is intended to go to `next` (prerelease) or to `main` (client v2 / redux v1)? 
Because the target branch is set as `next`, but the milestone is `dev-main-28-11-2023`

**Comment by dnamorim** [comment]
This PR is to be released on `next` (prerelease) or on `main` (v2 client / v1 redux)?

Because the target branch is set as `next`, but the milestone is set as `dev-main-28-11-2023`

**Review by ivomota — APPROVED** [review]

**Comment by hichico** [comment]
> This PR is to be released on `next` (prerelease) or on `main` (v2 client / v1 redux)?
> 
> Because the target branch is set as `next`, but the milestone is set as `dev-main-28-11-2023`

Hi @dnamorim 👋 
This branch is aiming for the next branch so we can ditch the appointed milestone.

File: packages/client/src/checkout/types/postCheckoutOrder.types.ts (hunk 4 above)
Comment (hichico) [reply packages/client/src/checkout/types/postCheckoutOrder.types.ts:14]: Waiting for response from the back-end team. Will answer you as soon as I have the right info.

**Pull request closed without merging**